pkg/sql/exec/coalescer.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package exec
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/sql/exec/types"
+)
+
+// coalescerOp consumes the input operator and coalesces the resulting batches
+// to return full batches of ColBatchSize.
+type coalescerOp struct {
+	input      Operator
+	inputTypes []types.T
+
+	group  ColBatch
+	buffer ColBatch
+}
+
+var _ Operator = &coalescerOp{}
+
+// NewCoalescerOp creates a new coalescer operator on the given input operator
+// with the given column types.
+func NewCoalescerOp(input Operator, colTypes []types.T) Operator {
+	return &coalescerOp{
+		input:      input,
+		inputTypes: colTypes,
+		buffer:     NewMemBatch(colTypes),
+	}
+}
+
+func (p *coalescerOp) Init() {
+	p.input.Init()
+}
+
+func (p *coalescerOp) Next() ColBatch {
+	p.group = p.buffer
+	p.buffer = NewMemBatch(p.inputTypes)
```

Comment:
Maybe this is why you got such bad benchmark results with this op, a while ago - this is allocating a whole new n columns every call to Next. It would be better to avoid this.

Reply:
Done. Oh yea this definitely made it faster. I also added a separate benchmark that doesn't wrap around any existing operators.